typing/typecore.ml

```diff
@@ -4032,15 +4035,30 @@ and type_expect_
               let error = Expr_not_a_record_type exp.exp_type in
               raise (Error (exp.exp_loc, env, error))
         in
+        (* In record update expressions [{r with ...}],
+           [r] may have an inline record type with a restricted scope.
+           We must propagate this restricted scope if we build a new
+           return type below. *)
+        let ty_record_scope = match opt_exp with
+          | None -> Ident.lowest_scope
+          | Some exp -> get_scope exp.exp_type
+        in
         match expected_opath, opt_exp_opath with
-        | None, None -> newvar (), None
+        | None, None ->
+            let ty = newvar () in
+            update_scope ty_record_scope ty;
```

Comment:
Why not use `new_scoped_ty` here ?
Could it happen that `!current_level < ty_record_scope` ?

Reply:
Thanks, I rewrote this part of the patch (and below) to use `new_scoped_ty`.

>  Could it happen that `!current_level < ty_record_scope` ?

That would mean that the `ty_record` type has escaped from its validity region, so there should be an error in this case, which is I think what we get here (but I don't really understand how/when the (scope <= level) condition is enforced in the type-checker).